### DISCUSSION TITLE
How to set different DataGridColumnHeader has different ContextMenu?

### DISCUSSION THREAD

**Discussion by nasterhe** [AvaloniaUI/Avalonia#17743] [Q&A] [answered]
#15252  provided a method, but the exact same ContextMenu will appear for all DataGridColumnHeaders.
```
<Style Selector="DataGridColumnHeader">
               <Setter Property="ContextMenu">
                  <ContextMenu>
                    <MenuItem Header="Test Header ContextMenu"/>                    
                  </ContextMenu>
                </Setter>                
              </Style>
```

Is there a way to set or not set the ContextMenu and set different ContextMenus for different headers? Thanks.

**Comment by msneijders** [comment]
Why not just set it in xaml like this?

Copypaste from ControlCatalog DataGridPage.xaml and adding a contextmenu to header:

```xaml
<DataGrid Name="dataGridGrouping" Margin="12">
    <DataGrid.Columns>
        <DataGridTextColumn
            Width="6*"
            x:DataType="local:Country"
            Binding="{Binding Name}"
            Header="Country or Region" />
        <DataGridTextColumn
            Width="4*"
            x:DataType="local:Country"
            Binding="{Binding Region}">
            <DataGridTextColumn.Header>
               <TextBlock Text="Region">
                  <TextBlock.ContextMenu>
                     <ContextMenu>
                        <MenuItem Header="Test Header ContextMenu" />
                    </ContextMenu>
                 </TextBlock.ContextMenu>
               </TextBlock>
            </DataGridTextColumn.Header>
        </DataGridTextColumn>
<!-- rest of datagrid -->
</DataGrid>
```

**Reply by msneijders** [reply, marked as answer]
Alternative way, if your header content is a constant:

```xaml
<DataGrid>
   <DataGrid.Styles>
     <Style Selector="DataGridColumnHeader[Content=Area]">
        <Setter Property="ContextMenu">
           <ContextMenu>
               <MenuItem Header="ContextMenu of column Area" />
            </ContextMenu>
         </Setter>
      </Style>
   </DataGrid.Styles>
</DataGrid>
```

Possible a custom handling of `Control.ContextRequestedEvent` could help as well -- but I do not see a reliable way of identifying which header is clicked, other then checking the header text, as most information about it is marked internal.

**Reply by nasterhe** [reply]
That's exactly what I wanted, thank you very much.

**Comment by nasterhe** [comment]
@msneijders 

![2024-12-24 173149](https://github.com/user-attachments/assets/a006d0b4-1bfb-4497-9d24-ee67be551471)

Thank you for your reply. As you can see, this effect makes the header wider and unsightly. And right-clicking on both sides of the header will not bring up the ContextMenu, instead of right-clicking anywhere in the DataGridColumnHeader, which will cause confusion to the user and think that there is something wrong.

**Reply by msneijders** [reply]
I updated my example above.

**Reply by nasterhe** [reply]
Thanks.
The method I'm using is the same as the one you provided, but this one has the disadvantage that the ContextMenu will only appear if you right-click on the text (i.e. TextBlock).
